accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,376 @@
+
+See also [C++\CLI documentation](https://learn.microsoft.com/en-us/cpp/extensions/interior-ptr-cpp-cli?view=msvc-170).
+
+An interior pointer is created in several ways:
+- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses the `ref byte` approach below.
+- Using `ref byte`. This is possible in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used. This `ref byte` approach is used today in reflection invoke when there are <=4 parameters.
+- Using `IntPtr` with tracking or pinning. Currently, tracking is only supported internally through the use of a "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.
+
+This design proposes the use of `TypedReference` for the safe version of APIs, and either `ref byte`, `IntPtr` or `TypeReference*` approaches for the unsafe version. The `ref byte` and `IntPtr` approaches require the use of pointers through the internal `object.GetRawData()` and other means which would need to be exposed publicaly or wrapped in some manner.
+
+## TypedReference
+The internal reflection implementation will not likely be based on `TypedReference` -- instead it will take, from public APIs, `TypedReference` instances and "peel off" the interior pointer either as a `ref byte` or as an `IntPtr` along with tracking to support GC. The use of `TypedReference` is essentially to not require the use of unsafe code in the public APIs. Unsafe public APIs will likely be added as well that should have slightly better performance characteristics.
+
+### Existing usages
+`TypedReference` is used in `FieldInfo` via [`SetValueDirect(TypeReference obj, object? value)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvaluedirect) and [`object? GetValueDirect(TypeReference obj)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.getvaluedirect). The approach taken by `FieldInfo` will be expanded upon in the design here. The existing approach supports the ability to get\set a field on a target value type without boxing. Boxing the target through [`FieldInfo.SetValue()`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvalue) is useless since the changes made to the boxed value type would not be reflected back to the original value type instance.
+
+`TypedReference` is also used by the undocumented `__arglist` along with `System.ArgIterator` although `__arglist` is Windows-only. The approach taken by `__arglist` will not be leveraged or expanded upon in this design. It would, however, allow a pseudo-strongly-typed approach like
```

Comment:
Should we add a top level item to track expanding `__arglist` to non-windows platforms? Most of other non-trivial work items have a top level item and my guess is this is non-trivial work.

Reply:
The idea of native varargs is tracked at https://github.com/dotnet/runtime/issues/48796.

Reply:
We should link to that here then.